File: frontend/src/lib/ioContracts.ts
/* Frontend mirror of the typed I/O contracts (I/O rebuild, Phase 1).

   These are the runtime-validated, versioned shapes for the input plane
   (InputSubmission — what the UI sends) and the operational-event plane
   (OutputEvent envelope + discriminated payload — what the UI receives). They
   mirror backend/agentflow/io_contracts.py. Validators are hand-written guards
   (consistent with lib/streamEvent.ts) so there is no new dependency; they never
   throw and reject unsupported schema versions, so malformed/foreign data fails
   safely instead of corrupting UI state. */

export const IO_SCHEMA_VERSION = "1";

// ---------------------------------------------------------------- input plane

export type InputReference =
  | { kind: "file"; path: string }
  | { kind: "folder"; path: string }
  | { kind: "diff"; path: string; staged?: boolean }
  | { kind: "task_artifact"; taskId: string; name: string }
  | { kind: "run"; runId: string }
  | { kind: "event_range"; fromEventId: number; toEventId: number };

export type InputDestination =
  | { kind: "controller" }
  | { kind: "provider"; provider: string }
  | {
      kind: "task";
      taskId: string;
      intent: "continue" | "clarify" | "retry" | "fix" | "reroute" | "ask";
    };

export type SubmitMode = "message" | "create_task" | "continue" | "retry" | "reroute";

export interface InputSubmission {
  schemaVersion: "1";
  id: string;
  workspaceId: string;
  destination: InputDestination;
  content: { text: string; references: InputReference[] };
  context: { taskId?: string; step?: string; provider?: string; orchestrationMode?: string };
  behavior: { submitMode: SubmitMode };
  createdAt: string;
}

/** Build a well-formed submission; the single way the UI expresses intent + destination. */
export function buildSubmission(args: {
  id: string;
  workspaceId: string;
  destination: InputDestination;
  text: string;
  references?: InputReference[];
  context?: InputSubmission["context"];
  submitMode?: SubmitMode;
  createdAt: string;
}): InputSubmission {
  return {
    schemaVersion: IO_SCHEMA_VERSION,
    id: args.id,
    workspaceId: args.workspaceId,
    destination: args.destination,
    content: { text: args.text, references: args.references ?? [] },
    context: args.context ?? {},
    behavior: { submitMode: args.submitMode ?? "message" },
    createdAt: args.createdAt,
  };
}

export function validateSubmission(raw: unknown): InputSubmission | null {
  if (!raw || typeof raw !== "object") return null;
  const r = raw as Record<string, unknown>;
  if (r.schemaVersion !== IO_SCHEMA_VERSION) return null; // unsupported version → fail safe
  const dest = r.destination as Record<string, unknown> | undefined;
  const content = r.content as Record<string, unknown> | undefined;
  if (!dest || typeof dest.kind !== "string") return null;
  if (!content || typeof content.text !== "string" || content.text.length === 0) return null;
  if (typeof r.id !== "string" || typeof r.workspaceId !== "string") return null;
  return raw as InputSubmission;
}

// ----------------------------------------------------- operational-event plane

export type EventPayload =
  | { type: "narrative.delta"; text: string }
  | { type: "narrative.completed"; text?: string }
  | { type: "command.started"; command: string }
  | { type: "command.output"; text: string }
  | { type: "command.completed"; exitCode?: number; durationMs?: number; status?: string }
  | { type: "task.state"; taskId: string; state: string }
  | { type: "queue.state"; activeCount?: number }
  | { type: "approval.requested"; approvalId: string; action: string; reason?: string }
  | { type: "approval.resolved"; approvalId: string; approved: boolean }
  | { type: "failure"; title: string; summary?: string }
  | { type: "cancellation"; runId?: string }
  | { type: "summary.ready"; kind: string };

const PAYLOAD_TYPES = new Set<string>([
  "narrative.delta",
  "narrative.completed",
  "command.started",
  "command.output",
  "command.completed",
  "task.state",
  "queue.state",
  "approval.requested",
  "approval.resolved",
  "failure",
  "cancellation",
  "summary.ready",
]);

export interface OutputEventEnvelope {
  schemaVersion: "1";
  id: string;
  workspaceId: string;
  createdAt: string;
  taskId?: string;
  runId?: string;
  channel?: "assistant" | "controller" | "stdout" | "stderr" | "system";
  sequence?: number;
  redacted: boolean;
  truncated: boolean;
  payload: EventPayload;
}

/** Validate a typed OutputEvent envelope; null on unsupported version or unknown
    payload type (fail-safe). */
export function validateOutputEvent(raw: unknown): OutputEventEnvelope | null {
  if (!raw || typeof raw !== "object") return null;
  const r = raw as Record<string, unknown>;
  if (r.schemaVersion !== IO_SCHEMA_VERSION) return null;
  if (typeof r.id !== "string" || typeof r.workspaceId !== "string") return null;
  const p = r.payload as Record<string, unknown> | undefined;
  if (!p || typeof p.type !== "string" || !PAYLOAD_TYPES.has(p.type)) return null;
  return raw as OutputEventEnvelope;
}

File: backend/agentflow/io_contracts.py
"""Typed input/output contracts (I/O rebuild, Phase 1).

Two planes are defined here as versioned, validated models:

- **Input plane** — :class:`InputSubmission`: one typed model for every input
  surface (controller chat, provider chat, task input). Destination and intent are
  explicit fields, never encoded inside free-form text.
- **Operational-event plane** — :class:`OutputEventEnvelope` + a discriminated
  ``payload`` union: the application-owned event contract. This replaces the open
  ``data: dict`` shape as the *authoritative* contract; ``event_bus`` continues to
  emit the legacy flat dict during migration, and :func:`validate_event` is the
  boundary validator the typed pipeline uses.

The deterministic-result plane (controller results + summaries) lives in
:mod:`agentflow.controller_protocol` and :mod:`agentflow.contracts`.
"""

from __future__ import annotations

from typing import Annotated, Literal, Optional, Union

from pydantic import BaseModel, Field, ValidationError

SCHEMA_VERSION = "1"


# ============================================================ INPUT PLANE


class FileReference(BaseModel):
    kind: Literal["file"] = "file"
    path: str


class FolderReference(BaseModel):
    kind: Literal["folder"] = "folder"
    path: str


class DiffReference(BaseModel):
    kind: Literal["diff"] = "diff"
    path: str
    staged: bool = False


class TaskArtifactReference(BaseModel):
    kind: Literal["task_artifact"] = "task_artifact"
    taskId: str
    name: str


class RunReference(BaseModel):
    kind: Literal["run"] = "run"
    runId: str


class EventRangeReference(BaseModel):
    kind: Literal["event_range"] = "event_range"
    fromEventId: int
    toEventId: int


InputReference = Annotated[
    Union[FileReference, FolderReference, DiffReference, TaskArtifactReference, RunReference, EventRangeReference],
    Field(discriminator="kind"),
]


class ControllerDestination(BaseModel):
    kind: Literal["controller"] = "controller"


class ProviderDestination(BaseModel):
    kind: Literal["provider"] = "provider"
    provider: str


class TaskDestination(BaseModel):
    kind: Literal["task"] = "task"
    taskId: str
    intent: Literal["continue", "clarify", "retry", "fix", "reroute", "ask"] = "continue"


InputDestination = Annotated[
    Union[ControllerDestination, ProviderDestination, TaskDestination],
    Field(discriminator="kind"),
]


class InputContent(BaseModel):
    text: str = Field(min_length=1, max_length=20_000)
    references: list[InputReference] = []


class InputContext(BaseModel):
    taskId: Optional[str] = None
    step: Optional[str] = None
    provider: Optional[str] = None
    orchestrationMode: Optional[str] = None


class InputBehavior(BaseModel):
    submitMode: Literal["message", "create_task", "continue", "retry", "reroute"] = "message"


class InputSubmission(BaseModel):
    schemaVersion: Literal["1"] = "1"
    id: str
    workspaceId: str
    destination: InputDestination
    content: InputContent
    context: InputContext = InputContext()
    behavior: InputBehavior = InputBehavior()
    createdAt: str


# ===================================================== OPERATIONAL EVENT PLANE
# Discriminated payloads, keyed by the envelope's ``type``. This is the typed
# contract that replaces the open ``data`` dict as the long-term shape.


class NarrativeDeltaPayload(BaseModel):
    type: Literal["narrative.delta"] = "narrative.delta"
    text: str


class NarrativeCompletedPayload(BaseModel):
    type: Literal["narrative.completed"] = "narrative.completed"
    text: str = ""


class CommandStartedPayload(BaseModel):
    type: Literal["command.started"] = "command.started"
    command: str


class CommandOutputPayload(BaseModel):
    type: Literal["command.output"] = "command.output"
    text: str


class CommandCompletedPayload(BaseModel):
    type: Literal["command.completed"] = "command.completed"
    exitCode: Optional[int] = None
    durationMs: Optional[int] = None
    status: Literal["succeeded", "failed", "cancelled", "error"] = "succeeded"


class TaskStatePayload(BaseModel):
    type: Literal["task.state"] = "task.state"
    taskId: str
    state: str


class QueueStatePayload(BaseModel):
    type: Literal["queue.state"] = "queue.state"
    activeCount: int = 0


class ApprovalRequestedPayload(BaseModel):
    type: Literal["approval.requested"] = "approval.requested"
    approvalId: str
    action: str
    reason: str = ""


class ApprovalResolvedPayload(BaseModel):
    type: Literal["approval.resolved"] = "approval.resolved"
    approvalId: str
    approved: bool


class FailurePayload(BaseModel):
    type: Literal["failure"] = "failure"
    title: str
    summary: str = ""


class CancellationPayload(BaseModel):
    type: Literal["cancellation"] = "cancellation"
    runId: Optional[str] = None


class SummaryReadyPayload(BaseModel):
    type: Literal["summary.ready"] = "summary.ready"
    kind: str  # a contracts.py summary kind (task_summary/test_summary/…)


EventPayload = Annotated[
    Union[
        NarrativeDeltaPayload,
        NarrativeCompletedPayload,
        CommandStartedPayload,
        CommandOutputPayload,
        CommandCompletedPayload,
        TaskStatePayload,
        QueueStatePayload,
        ApprovalRequestedPayload,
        ApprovalResolvedPayload,
        FailurePayload,
        CancellationPayload,
        SummaryReadyPayload,
    ],
    Field(discriminator="type"),
]


class OutputEventEnvelope(BaseModel):
    schemaVersion: Literal["1"] = "1"
    id: str
    workspaceId: str
    createdAt: str
    taskId: Optional[str] = None
    runId: Optional[str] = None
    chatId: Optional[str] = None
    messageId: Optional[str] = None
    queueItemId: Optional[str] = None
    approvalId: Optional[str] = None
    provider: Optional[str] = None
    step: Optional[str] = None
    channel: Optional[Literal["assistant", "controller", "stdout", "stderr", "system"]] = None
    sequence: Optional[int] = None
    redacted: bool = False
    truncated: bool = False
    payload: EventPayload


def validate_event(data: dict) -> tuple[Optional[OutputEventEnvelope], Optional[str]]:
    """Validate a raw event dict against the typed envelope. Returns
    ``(envelope, None)`` or ``(None, error_message)`` — never raises."""
    try:
        return OutputEventEnvelope.model_validate(data), None
    except ValidationError as exc:
        return None, f"{exc.error_count()} validation error(s)"


def validate_submission(data: dict) -> tuple[Optional[InputSubmission], Optional[str]]:
    """Validate a raw input submission. Returns ``(submission, None)`` or
    ``(None, error_message)`` — never raises."""
    try:
        return InputSubmission.model_validate(data), None
    except ValidationError as exc:
        return None, f"{exc.error_count()} validation error(s)"
